src/v/utils/external_process.cc

```diff
@@ -0,0 +1,162 @@
+  std::filesystem::path path,
+  ss::experimental::spawn_parameters params)
+  : _process(std::move(p))
+  , _path(std::move(path))
+  , _params(std::move(params))
+  , _terminate_mutex("external_process::terminate")
+  , _wait_mutex("external_process::wait") {}
+
+external_process::external_process(external_process&& rhs) noexcept
+  : _process(std::move(rhs._process))
+  , _path(std::move(rhs._path))
+  , _params(std::move(rhs._params))
+  , _completed(rhs._completed)
+  , _terminate_mutex(std::move(rhs._terminate_mutex))
+  , _wait_mutex(std::move(rhs._wait_mutex))
+  , _stdout_consumer(std::exchange(rhs._stdout_consumer, std::nullopt))
+  , _stderr_consumer(std::exchange(rhs._stderr_consumer, std::nullopt))
+  , _gate(std::move(rhs._gate)) {
+    rhs._completed = true;
+}
+
+external_process::~external_process() {
+    vassert(_completed, "Process has not yet completed");
+    vassert(!_stdout_consumer.has_value(), "stdout consumer not cleaned up");
+    vassert(!_stderr_consumer.has_value(), "stderr consumer not cleaned up");
+}
+
+ss::future<ss::experimental::process::wait_status> external_process::wait() {
+    auto guard = _gate.hold();
+    if (_completed) {
+        throw std::system_error(
+          make_error_code(error_code::process_already_completed));
+    }
+
+    auto units = _wait_mutex.try_get_units();
+    if (!units) {
+        throw std::system_error(
+          make_error_code(error_code::process_already_being_waited_on));
+    }
+
+    auto resp = co_await _process.wait();
+
+    if (_stdout_consumer.has_value()) {
+        try {
+            co_await std::move(_stdout_consumer).value();
+        } catch (...) {
+            // ignore the error if `input_stream::consume` throws
+        }
+
+        _stdout_consumer.reset();
+    }
+
+    if (_stderr_consumer.has_value()) {
+        try {
+            co_await std::move(_stderr_consumer).value();
+        } catch (...) {
+            // ignore the error if `input_stream::consume` throws
+        }
+
+        _stderr_consumer.reset();
+    }
+
+    _completed = true;
+
+    co_return resp;
+}
+
+ss::future<> external_process::terminate(std::chrono::milliseconds timeout) {
+    auto guard = _gate.hold();
+
+    if (_completed) {
+        throw std::system_error(
+          make_error_code(error_code::process_already_completed));
+    }
+
+    auto units = _terminate_mutex.try_get_units();
+    if (!units) {
+        throw std::system_error(
+          make_error_code(error_code::process_already_being_terminated));
+    }
+
+    if (_wait_mutex.ready()) {
+        throw std::system_error(
+          make_error_code(error_code::terminate_called_before_wait));
+    }
+
+    try {
+        _process.terminate();
+    } catch (const std::exception& e) {
+        throw std::system_error(
+          make_error_code(error_code::failed_to_terminate_process), e.what());
+    }
+
+    bool cancel = false;
+    return ss::do_with(cancel, [this, timeout](bool& cancel) {
```

Comment:
Do we need to hold the units here? Why not use coroutines here?

Reply:
Can you please expand on what you mean here.

Reply:
There is a lock that is being held above - that is only held synchronously and not doing anything during this async work.

If we do need to keep the units alive right now, then let's use coroutines instead of this do_with?

Reply:
> There is a lock that is being held above - that is only held synchronously and not doing anything during this async work.

Thanks good catch

Sorry by units I thought for some reason you meant the timeout value not the variable `units`, my bad.

> let's use coroutines instead of this do_with

The issue is that I have a loop checking to see if the `_completed` flag goes true so the `ss::with_timeout` can exit.  My understanding is that even if the `with_timeout` times out, the future (which in this case is a `do_until`) doesn't get cancelled.

Reply:
big thanks to @oleiman rubber 🦆 'ing with me, found I think a better solution